microsoft/MixedRealityToolkit-Unity
# Unified input methods and new features for input simulation

## Overview

Input simulation camera control and hand control originated from two separate features and have inconsistent support for rotation as well as slightly different approaches to key binding.

This PR aims to unify the support for keyboard/mouse/game controller input and complement existing features.

## Changes
- Fixes: #5489 
- New  editor tools window to expose input simulation state of head and hands at runtime. This also provides useful features such as "Reset" buttons to bring head/hands back to default state. This window also incorporates the previous "Input Recording" window.
- Optional viewport icons for controlling hands, similar to icons in HTK.
- Introduces a generic `KeyBinding` struct that encodes both KeyCode and mouse button. This comes with editor support for quickly assigning buttons with a key press. (_Note: Unity editor does not support shift key presses on their own, so these have to be assigned using the legacy enum popup).
- Unifies handling of mouse rotations for camera and hands, including cursor locking and visibility. The previous 6 keys for rotating hands have been removed in favor of intuitive mouse rotation and/or changing angles in the editor window.
- Support for roll angle of head and hands, using the scroll wheel during rotation (pitch and yaw are controlled by mouse x/y movement as before).
- New input axis "UpDown" to control camera movement along the Up vector. Defaults to Q/E keys and the 3rd game controller axis (triggers)

**Breaking Changes:** InputSimulationProfile has changed, with some existing KeyCode fields replaced by the new KeyBinding struct (this might reset user profiles to default settings). Some fields have been renamed, but should be handled by FormerlySerializedAs attributes.

## Verification
- Read through input simulation and input animation recording documentation and check for correctness and clarity.
- Run the hand tracking test scene in the editor and try out input simulation features:
  * Camera movement using WASD and game controller
  * Camera rotation by holding RMB
  * Hand movement using LShift/Space + mouse movement
  * Hand rotation using LShift/Space + RMB + mouse movement (it becomes less complicated when toggling hands on using T/Y first)
  * Try camera/hand roll by using the mouse wheel while rotating
  * Move camera up/down using Q/E keys or game controller triggers
- Open the Input Simulation window and try buttons
- Also try assigning new key bindings using the little buttons on the right.
- Try toggling hands via the icons in the viewport in play mode.
